### DISCUSSION TITLE
Transactions starting at unexpected times

### DISCUSSION THREAD

**Discussion by LachlanMarnham** [sqlalchemy/sqlalchemy#10941] [Usage Questions] [answered]
Hello. I am trying to use SQLAlchemy to write `SELECT ... AS OF ...` queries against a SQL Server [system-versioned table](https://learn.microsoft.com/en-us/sql/relational-databases/tables/temporal-tables?view=sql-server-ver16). Basically, system-versioned tables version the data with timestamps so you can do something like

`SELECT * FROM my_table FOR SYSTEM_TIME AS OF <timestamp>` 

and get back everything which was in the table at `<timestamp>`.

I have written the following example of the problem I'm facing:

 ```python
from datetime import datetime
from time import sleep

from sqlalchemy import (
    create_engine,
    text,
)


url = (
    "mssql+pyodbc://core_test_user:TestUser1!@127.0.0.1:1433/taco_core_test"
    "?TrustServerCertificate=yes&driver=ODBC+Driver+18+for+SQL+Server"
)

engine = create_engine(url, echo=True)

row_1 = {"col_1": "1", "col_2": 1}
row_2 = {"col_1": "2", "col_2": 2}

pre_inserts = datetime.utcnow()
sleep(0.01)
with engine.begin() as conn:
    conn.execute(text(
        "INSERT INTO portfolio.versioned_table (col_1, col_2) VALUES ('1', 1);"
    ))

sleep(0.01)
post_insert_row_1 = datetime.utcnow()

sleep(0.01)
with engine.begin() as conn:
    conn.execute(text(
        "INSERT INTO portfolio.versioned_table (col_1, col_2) VALUES ('2', 2);"
    ))
sleep(0.01)
post_insert_row_2 = datetime.utcnow()


with engine.begin() as conn:
    result_1 = conn.execute(text(
        f"SELECT * FROM portfolio.versioned_table FOR SYSTEM_TIME AS OF  '{pre_inserts}';"
    )).fetchall()
    result_2 = conn.execute(text(
        f"SELECT * FROM portfolio.versioned_table FOR SYSTEM_TIME AS OF  '{post_insert_row_1}';"
    )).fetchall()
    result_3 = conn.execute(text(
        f"SELECT * FROM portfolio.versioned_table FOR SYSTEM_TIME AS OF  '{post_insert_row_2}';"
    )).fetchall()

# result_1 == []
# result_2 == [('1', 1), ('2', 2)], expected to get [('1', 1)]
# result_3 == [('1', 1), ('2', 2)]
```

Basically I'm checking what the time is after each `INSERT`, and using that later in a `SELECT ... AS OF ` query, but `result_2` tells me there was an extra row in the database before that row was ever inserted.

SQL Server versions the data according to the time at which the transaction to insert a given row was started. But when I look at the server, it thinks the transaction was started before my call to `engine.begin`.

Things I've tried:
1. `isolation_level="AUTOCOMMIT"` fixes the problem, but I can't just switch transactions off for our whole codebase
2. Confirmed it's not a problem with `pyodbc` (the problem also occurs with `pymssql`)
3. Confirmed it's not a problem with SQL Server (everything works as expected when manually executing commands against the database with a GUI client)
4. Confirmed this isn't just that the clocks on client and server side are a bit different. The `sleep`s above can be 0.01s or 100s, and the behaviour is always the same: the time when SQL Server thinks the first row was inserted is always about the same as the time it thinks the second row was inserted

My vague hypothesis is that the transaction in which row1 was inserted is somehow being reused. It's very strange that, no matter how long passes between the two rows being inserted, SQL Server always thinks the transaction for the second row was immediately after the transaction for the first.

I am a bit wary about what [this part of the SQLAlchemy docs](https://docs.sqlalchemy.org/en/20/tutorial/dbapi_transactions.html) means:

> You might have noticed the log line “BEGIN (implicit)” at the start of a transaction block. “implicit” here means that SQLAlchemy did not actually send any command to the database; it just considers this to be the start of the DBAPI’s implicit transaction.

But for system-versioned tables the exact time at which the transaction started is very important. Is there a way to change this behaviour? Or does anybody have any other insight? I'd be very grateful thanks.

**Comment by zzzeek** [comment]
this is very basic SQL execution code, so perhaps try with the pyodbc driver directly instead, and see if the same problems exist.  the "implicit BEGIN" thing is done by the DBAPIs (drivers, in this case pyodbc), so if that's a problem, you may want to emit the word "BEGIN" as a statement by itself, however pyodbc would hopefully integrate with that.

**Reply by LachlanMarnham** [reply]
Hi @zzzeek many thanks for the suggestion (and for SQLAlchemy). I think I have made some progress with this now:

**What the problem actually was**
`pyodbc` is opening transactions as soon as soon as the previous transaction is committed. I made the following modification to the code above:

```python
with engine.begin() as conn:
   conn.execute(text(
       "INSERT INTO portfolio.versioned_table (col_1, col_2) VALUES ('1', 1);"
   ))
   pre_commit = datetime.utcnow()
post_commit = datetime.utcnow()
``` 
and was able to see that the `valid_from` time for the *next* row was in between `pre_commit` and `post_commit`. In other words: we INSERT the first row, end our transaction, a new transaction is immediately started and that start time is then associated with the INSERT of the second row. I can't tell if this is unexpected or buggy behaviour, but it's definitely worrisome for anyone working with system versioned tables: a long-running process which writes to a database once per day would have timestamps in the `valid_from` column which were incorrect by a full day.
The `pyodbc` [docs](https://github.com/mkleehammer/pyodbc/wiki/Database-Transaction-Management) say this:

> When using pyodbc with autocommit=False, it is important to understand that you never explicitly open a database transaction in your Python code. Instead, a database transaction is implicitly opened when a Connection object is created with pyodbc.connect().

But I am using `with engine.connect` to start my transactions, so apparently `pyodbc.connect()` is called pre-emptively so there is a `pyodbc.Connection` ready to go the next time `sqlalchemy` creates a `sqlalchemy.Connection`.

As I mentioned above, we also saw this behaviour with `pymssql`, which effectively makes system versioning incompatible with both drivers without some extra work IMO.

**The fix**
setting the `isolation_level` to `"AUTOCOMMIT"` does fix the versioning issue, but we lose transactions. Therefore, I set it to `"AUTOCOMMIT"` at `Engine`-instantiation time, and then re-set the `isolation_level` to `"READ COMMITTED"` every time we get a new connection.  It basically looks like this:

```python
def _begin_event_handler(conn: Connection) -> None:
    conn.execution_options(isolation_level="READ COMMITTED")

def engine_factory(url):
    engine = sqlalchemy.create_engine(url, isolation_level="AUTOCOMMIT")
    event.listen(engine, "begin", _begin_event_handler)
```

**Reply by zzzeek** [reply]
you should be able to work the other way around, use a normal engine and then when you need this particular series of operations, set `isolation_level='AUTOCOMMIT'` on just the `Connection` you're working with.

**Reply by LachlanMarnham** [reply]
If I am not mistaken, that would mean:

```
with engine.begin() as conn:
    conn.execute(...)
    conn.execute(...)
```
is two transactions, not one. If the second `execute` failed, the first would not be rolled back.

**Reply by zzzeek** [reply]
oh you have an event handler, I see.   I missed the listen() part.   setting up READ COMMITTED like that emits the BEGIN?   OK then.  usually we just tell people to emit the SQL "BEGIN".

**Reply by CaselIT** [reply, marked as answer]
This is probably due to the queue pool used by default, since a connection will stay in the pool.
I think possible other solutions could be
- use a null pool, so that every time you call `engine.connect()` a new connection is created
- add a checkhout event to call rollback every time a connection is checked out from the pool https://docs.sqlalchemy.org/en/20/core/events.html#sqlalchemy.events.PoolEvents.checkout
- maybe pool_pre_ping will basically do the same as the event to call rollback mentioned above?

@zzzeek is there an option similar to https://docs.sqlalchemy.org/en/20/core/pooling.html#sqlalchemy.pool.Pool.params.reset_on_return but that controls what the pool does on checkout?

**Reply by zzzeek** [reply]
the pool doesnt do anything on checkout, if you want to do something, use the checkout event

**Reply by LachlanMarnham** [reply]
Ah very interesting. So it's not that the next transaction is started automatically when the last one ends, it just happens to work out that way because I've been testing with a number of inserts which is less than the pool size. If the number of insert statements were larger than the pool size the behaviour may change once connections start being resused. Unless this is actually true:

```
setting up READ COMMITTED like that emits the BEGIN? OK then. 
```

I'll check.

**Reply by CaselIT** [reply]
that's what I'm assuming, if the driver is always emitting a `BEGIN` after a `ROLLBACK` to start a new transaction that would be the behaviour that you would see.
Testing if it's the case should be easy enough using the event
